WIP: Feat/openrpc describe

meowsbits opened the pull request:
Rel #63 

This is a spike to explore runtime generation of an OpenRPC `rpc_discover` document that completely describes geth's available RPC services. Working initially to pattern this closely against geth, and then abstract to a pluggable package that can be integrated with go-ethereum/rpc, and maybe also expose other OpenRPC library services.

Provided temporarily (for development only) in this PR are a script and logs, as well as a copy of the openrpc.json document itself. These together are intended to make changes to the openrpc.json document transparent and to give context for their implementation. The `jj` the script requires is https://github.com/tidwall/jj. 

Included as submodules and forked for development are cornerstone libraries `github.com/alecthomas/jsonschema` and `github.com/gregdhill/go-openrpc`. 

An example of what adding a method looks like: https://github.com/etclabscore/core-geth/commit/7cbfe11432147eb1d6c412bbe997c7ed2137a76a

### Yields

:page_facing_up: `rpc.discover` document: https://gist.github.com/meowsbits/4da4c08765679dac1899543002d1f545. Note that geth uses a temporary `rpc_describe` method in lieu of the eventual `rpc_discover` name. This is in order to avoid the collision with the exising method, which returns the static **ethereum-json-rpc-specification** document.

:eye: Check it out on the **OpenRPC Playground**: https://playground.open-rpc.org/?url=https://gist.githubusercontent.com/meowsbits/4da4c08765679dac1899543002d1f545/raw/spec.json

### Develop

```
> ./.develop/test.sh
```
> https://github.com/etclabscore/core-geth/blob/feat/openrpc-describe/.develop/test.sh

meowsbits committed:
rpc: a way to describe openrpc doc vs. actual api methods

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: return more information abound under methods

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: change method to dedupe, and instead read from file

It's annoying to managing passing the contents of a file
as an item in an array from the command line.
And the document has strange escape and special characters.
So this make that a little less hackrequire.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
openrpc.json: include current ethereum-json-rpc-specification document in root

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
: include script for developing openrpc document

This is just a quick little developer tool.
It'll produce some logs which we can include in the
development branch as well. I hope this'll help with
transparency of workflow and easier review.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
: initial results for openrpc development feedback tools

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
: use pretty printing and sorted Over lists

This should make it a lot easier to actually read.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
: revise further with hopefully improved/simplified alphabetization scheme

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
openrpc: remove eth_submitHashrate

It was in the Under list, ergo not available from the
actual server. Was listed as deprecated. Has met its fate.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
openrpc: add admin_addPeer

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
openrpc: add admin_addTrustedPeer

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
openrpc: add admin_datadir

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
develop-openrpc-doc.sh: exit early if openrpc doc as edited isn't valid

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
WIP: rpc api describing with runtime, ast

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: getting closer to openrpc schema for under list

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
:hammer:

meowsbits committed:
WIP: tests and refactoring Describe

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
install alecthomas.jsonschema for cd schemas

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:floppy_disk:

meowsbits committed:
:floppy_disk:

meowsbits committed:
WIP: was wrestling with nil-as-oneof but i'm buggy

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
wip: rpc_describe returns doc

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
run cat rpc_describe.log | jj -p 'result' > ethereum-json-rpc.json

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
use external links instead of description

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
add ServerWithListener fn so server can self-describe

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
server description from provider program name, version

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
experiments with ethapi structs, fixes TraceTransaction return type

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/openrpc-validator-cli.js,.develop/spec.json,.develop/test.sh,.gitmodules,go-openrpc,go.mod,p2p,rpc: set up reproducable development environment, tests

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
p2p,rpc: (lint) goimports

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.gitmodules: make go-openrpc development version remote

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/test.sh: fixup test script

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/test.sh: fixup test script

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/test.sh: fixup test script

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
:hammer:

meowsbits committed:
.develop/spec.json,.develop/spec1.json,.develop/spec2.json,rpc: passes validator. still might be wrong.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
:floppy_disk:

meowsbits committed:
.develop/spec2.json,rpc: maybe have deferencing too

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,.develop/spec2.json,rpc: closer than ever

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: remove dead code and dead comments

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go-openrpc: update go-openrpc submodule

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,.develop/spec2.json,rpc: :bookmark: it passes

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,ethapi,rpc: messing around with EthSyncing, trying to figure oneOf out

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: fix typo

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.gitmodules,jsonschema: develop fork of jsonschema lib

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/serve.py,.develop/spec.json,go.mod,ethapi,jsonschema,rpc: still working eth_syncing and oneof issues

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,rpc: re-enable 'cleaned' openrpc document

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
ethapi,rpc: (lint) goimports, govet, gosimple

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: update test callbacks count to actual

I introduced some new methods for additional inspection.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec2.json: update test output

Signed-off-by: meows <b5c6@protonmail.com>

BelfordZ commented on .develop/openrpc-validator-cli.js:
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env node
+"use strict";
+Object.defineProperty(exports, "__esModule", { value: true });
```
is this the result of running typescript compile?

If so, might I suggest that you include the src ts instead of the compiled result?

If this is meant to be maintained as JS, then lets ditch the export and clean this up a little.

meowsbits replied:
Noted. This is/was only in the PR as a shortcut/convenience for reproducing the document validation heuristic, and I've since made [the script just demand that this tool is installed globally](https://github.com/etclabscore/core-geth/blob/feat/openrpc-describe/.develop/test.sh#L42).

BelfordZ commented on .develop/serve.py:
```diff
@@ -0,0 +1,61 @@
+import SimpleHTTPServer
```
why add the python dep?

heres a popular one: https://www.npmjs.com/package/serve

meowsbits replied:
This is here so I can serve the generated doc to a local Open-RPC playground for development visibility. I was getting CORS errors with the normal `SimpleHTTPServer` on Firefox otherwise.

BelfordZ commented on .develop/spec2.json:
```diff
@@ -0,0 +1,5841 @@
+{
```
Recommended that you use @open-rpc/examples here instead - this way you will not have to update your tests to know what version of OpenRPC your code works / doesnt work for.

meowsbits replied:
`spec2.json` is a file generated by the tests.

Well noted on the .@examples point, and will do when I start pulling the logic out and building   tests for it in isolation.

BelfordZ commented on internal/ethapi/api.go:
```diff
@@ -73,27 +74,56 @@ func (s *PublicEthereumAPI) ProtocolVersion() hexutil.Uint {
+//
+// FIXME: Can't get jsonschema oneOf to work right.
+//
+type EthSyncingResult struct {
```
Perhaps you can infer this from the contents of the struct?

for example, when you encounter a struct, it  can be any of the following json schema types:
1. object
2. allOf
3. anyOf

As it turns out, any of these can be in effect, just an object. However, you figure out which to use based on:
- if all the keys in the struct are optional, it MAY be an 'anyOf'
- If all the keys in the struct are required, it MAY be an 'allOf'
- if there is some keys required, others not, it MUST be an object.

While following these heuristics only wont give a perfect solution, perhaps its worth a try?

meowsbits committed:
rpc: add test for cyclical graph, fail

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
:floppy_disk:

meowsbits committed:
:hammer:

meowsbits committed:
:hammer:

meowsbits committed:
:hammer:

meowsbits committed:
:hammer:

meowsbits committed:
jst: clean up

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod,go.sum,jst: initial mocking for TestTraverse

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jst: logs, set tests to pass

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jst: debugging, refactor Traverse to look like the typescript orig

The refactoring of traverse to use the closure
doesn't effect test outcomes.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jst: refactor Traverse test cases to use mutate/expect design

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
:floppy_disk:

meowsbits committed:
jst: clean up dead and commented code

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:floppy_disk:

meowsbits committed:
:hammer:

meowsbits committed:
jst,rpc: abstract test traverse options logic

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
i actually think uniquenodes option should not be an option

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:floppy_disk:

meowsbits committed:
jst: notes, clean

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
go.mod,go.sum: go mod tidy

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jst: few more tests and logs, pretty stable

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jst,rpc: init refactoring

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,ethapi,jst,rpc: continue cleanup, fixes oneof issue in jsonschema

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod,go.sum: maybe go mod tidy

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,scwallet,jsonschema,rpc: hacky workarounds for misbehaving reflector oneofs

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jsonschema: bump jsonschema to latest

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,rpc: run test

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jst,rpc: (lint) goimports

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,jst,rpc: fixes rpc.BlockNumberOrHash type...

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
jsonschema: bump jsonschema

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,node,openrpc,rpc,rpc: init big refactoring to make openrpc generic, so far so good

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.sum,openrpc,rpc: refactoring openrpc stuff out

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod,go.sum: go mod updates for removed jsonschema-traverse package

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
:hammer:

meowsbits committed:
.gitmodules,go-openrpc: remote go-openrpc as submodule, we'll use as fork instead

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod: update to use fork

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.sum,openrpc,rpc: begin reorg

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: checkout foundation/master -- rpc

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main,go.mod,go.sum,graphql,node,openrpc,rpc: chechout node, graphql, rpc foundation/master

There was a refactoring of node/rpc handler.
This updates to that, and isolates openrpc logic
to openrpc, where it's waiting to be implemented

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.gitmodules,jsonschema: remove extraced submodule jsonschema

Now using as a fork

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod,go.sum: update jsonschema fork imports

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
openrpc,rpc: init openrpc discover impl

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,.develop/test.sh,go.mod,rpc: init working impl

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
: remove disused openrpc static files

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: update to new rpc/node pattern

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
rpc: (lint:deadcode) remote newGzipHandler

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
ethapi: lint:goimports

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.golangci.yml,rpc: lint:deadcode add rpc case to ignored issues

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod,go.sum: bump openrpc dependency versions

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Merge branch 'master' into feat/openrpc-describe

Conflicts:
	cmd/geth/retesteth.go
	go.mod
	go.sum
	rpc/http.go

meowsbits committed:
.develop/spec.json,go.mod,go.sum: update go-openrpc-reflect, improving contentdescriptor naming

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,rpc: provide type to schema map for more hexutil types

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,go.mod,go.sum: bump go-openrpc-reflect dependency v0.0.19

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,.develop/test.sh,rpc: register rpc.discover method under rpc module namespace

It was under 'openrpc' for development to avoid
side effects of collision with the existing methods under
that module.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,node,rpc: refactor openrpc registry to node/ (out of rpc/)

The openrpc service shouldn't be installed in the
rpc library itself, since the library should only provide
the actual rpc service. Service registration, as it were,
should be handled by the layer using the rpc service,
in this case node/.
L

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main,node: move openrpc descriptor service to each handler

This allows discrete ipc/http/ws module sets, each
with their own rpc.discover document based on what
is actually being handled.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
.develop/spec.json,go.mod,node: remove redundant zeroed instances in jsonschema typemapper

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
node,params: lint:goimports

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
go.mod,go.sum: bump go-openrpc-reflect v0.0.20

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
node: fix node stack name -> Core-Geth

This was overwritten when I checked out the
node and rpc packages from foundation/master.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
bind: feature to keep the bind test 'replace' dependencies 1:1 with root go.mod

Signed-off-by: meows <b5c6@protonmail.com>
